File: demo-app-objc/CloudXObjCRemotePods/Views/Ad/NativeFeedViewController.m
#import "NativeFeedViewController.h"
#import <CloudXCore/CloudXCore.h>
#import "CLXDemoConfigManager.h"
#import "DemoAppLogger.h"
#import "LogsModalViewController.h"
#import "UserDefaultsSettings.h"

static const NSInteger kAdCount = 10;
static NSString * const kAdCellIdentifier = @"NativeAdCell";

typedef NS_ENUM(NSInteger, NativeAdCreativeType) {
    NativeAdCreativeTypeUnknown,
    NativeAdCreativeTypeImage,
    NativeAdCreativeTypeVideoLandscape,
    NativeAdCreativeTypeVideoPortrait,
};

#pragma mark - NativeAdFeedItem

@interface NativeAdFeedItem : NSObject
@property (nonatomic, strong) CLXNativeAdLoader *loader;
@property (nonatomic, strong, nullable) CLXNativeAdView *adView;
@property (nonatomic, strong, nullable) CLXAd *ad;
@property (nonatomic, assign) NativeAdCreativeType creativeType;
@property (nonatomic, assign) BOOL isLoading;
@property (nonatomic, assign) BOOL isLoaded;
@property (nonatomic, assign) BOOL isFailed;
@end

@implementation NativeAdFeedItem
@end

#pragma mark - NativeFeedViewController

@interface NativeFeedViewController () <UITableViewDataSource, UITableViewDelegate, CLXNativeAdDelegate, CLXAdRevenueDelegate>
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) NSMutableArray<NativeAdFeedItem *> *feedItems;
@property (nonatomic, assign) NSInteger nextLoadIndex;
@end

@implementation NativeFeedViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Native Feed";
    self.view.backgroundColor = [UIColor systemBackgroundColor];

    [self setupTableView];
    [self setupAppLogsButton];
    [self createFeedItems];
    [self loadNextAd];
}

- (void)dealloc {
    for (NativeAdFeedItem *item in self.feedItems) {
        [item.loader destroy];
    }
}

#pragma mark - Setup

- (void)setupTableView {
    self.tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.allowsSelection = NO;
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:kAdCellIdentifier];
    [self.view addSubview:self.tableView];

    [NSLayoutConstraint activateConstraints:@[
        [self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
        [self.tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self.tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
    ]];
}

- (void)setupAppLogsButton {
    UIButton *logsButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [logsButton setTitle:@"App Logs" forState:UIControlStateNormal];
    logsButton.titleLabel.font = [UIFont systemFontOfSize:14 weight:UIFontWeightMedium];
    logsButton.backgroundColor = [UIColor systemOrangeColor];
    [logsButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    logsButton.layer.cornerRadius = 6;
    logsButton.translatesAutoresizingMaskIntoConstraints = NO;
    [logsButton addTarget:self action:@selector(showLogsModal) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:logsButton];

    [NSLayoutConstraint activateConstraints:@[
        [logsButton.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:8],
        [logsButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-12],
        [logsButton.widthAnchor constraintEqualToConstant:90],
        [logsButton.heightAnchor constraintEqualToConstant:30],
    ]];
}

- (void)showLogsModal {
    LogsModalViewController *logsModal = [[LogsModalViewController alloc] initWithTitle:@"Logs"];
    logsModal.modalPresentationStyle = UIModalPresentationPageSheet;
    [self presentViewController:logsModal animated:YES completion:nil];
}

- (NSString *)adUnitId {
    NSString *adUnitId = [[CLXDemoConfigManager sharedManager] currentConfig].nativeAdUnitId;
    if ([UserDefaultsSettings sharedSettings].nativeMediumAdUnitId.length > 0) {
        adUnitId = [UserDefaultsSettings sharedSettings].nativeMediumAdUnitId;
    }
    return adUnitId;
}

- (void)createFeedItems {
    self.feedItems = [NSMutableArray arrayWithCapacity:kAdCount];
    self.nextLoadIndex = 0;

    for (NSInteger i = 0; i < kAdCount; i++) {
        NativeAdFeedItem *item = [[NativeAdFeedItem alloc] init];
        item.loader = [[CloudXCore shared] createNativeAdLoaderWithAdUnitIdentifier:[self adUnitId]];
        item.loader.nativeAdDelegate = self;
        item.loader.revenueDelegate = self;
        item.loader.placement = [NSString stringWithFormat:@"feed_slot_%ld", (long)i];
        item.creativeType = NativeAdCreativeTypeUnknown;
        [self.feedItems addObject:item];
    }
}

#pragma mark - Sequential Loading

- (void)loadNextAd {
    if (self.nextLoadIndex >= kAdCount) return;

    NativeAdFeedItem *item = self.feedItems[self.nextLoadIndex];
    item.isLoading = YES;
    [item.loader loadAd];
}

- (nullable NativeAdFeedItem *)feedItemForLoader:(CLXNativeAdLoader *)loader {
    for (NativeAdFeedItem *item in self.feedItems) {
        if (item.loader == loader) return item;
    }
    return nil;
}

- (NSInteger)indexForLoader:(CLXNativeAdLoader *)loader {
    for (NSInteger i = 0; i < self.feedItems.count; i++) {
        if (self.feedItems[i].loader == loader) return i;
    }
    return NSNotFound;
}

#pragma mark - Creative Type Detection

- (NativeAdCreativeType)detectCreativeTypeFromAd:(CLXAd *)ad nativeAdView:(CLXNativeAdView *)adView {
    CLXNativeAd *nativeAd = ad.nativeAd;
    if (!nativeAd) return NativeAdCreativeTypeUnknown;

    if (nativeAd.isVideoContent) {
        CGFloat ar = nativeAd.mediaContentAspectRatio;
        return (ar < 1.0) ? NativeAdCreativeTypeVideoPortrait : NativeAdCreativeTypeVideoLandscape;
    }

    return (nativeAd.mediaContentAspectRatio > 0) ? NativeAdCreativeTypeImage : NativeAdCreativeTypeUnknown;
}

- (NSString *)badgeTextForCreativeType:(NativeAdCreativeType)type {
    switch (type) {
        case NativeAdCreativeTypeImage: return @"IMAGE";
        case NativeAdCreativeTypeVideoLandscape: return @"VIDEO 16:9";
        case NativeAdCreativeTypeVideoPortrait: return @"REELS 9:16";
        case NativeAdCreativeTypeUnknown: return @"UNKNOWN";
    }
}

- (UIColor *)badgeColorForCreativeType:(NativeAdCreativeType)type {
    switch (type) {
        case NativeAdCreativeTypeImage: return [UIColor systemBlueColor];
        case NativeAdCreativeTypeVideoLandscape: return [UIColor systemOrangeColor];
        case NativeAdCreativeTypeVideoPortrait: return [UIColor systemPurpleColor];
        case NativeAdCreativeTypeUnknown: return [UIColor systemGrayColor];
    }
}

#pragma mark - CLXNativeAdDelegate

- (void)didLoadNativeAd:(nullable CLXNativeAdView *)nativeAdView forAd:(CLXAd *)ad {
    CLXNativeAdLoader *loader = nil;
    for (NativeAdFeedItem *item in self.feedItems) {
        if (item.isLoading && !item.isLoaded) {
            loader = item.loader;
            break;
        }
    }

    NSInteger idx = self.nextLoadIndex;
    if (idx >= kAdCount) return;

    NativeAdFeedItem *item = self.feedItems[idx];
    item.isLoading = NO;
    item.isLoaded = YES;
    item.ad = ad;

    if (nativeAdView) {
        item.adView = nativeAdView;
    } else {
        CLXNativeAdView *templateView = [CLXNativeAdView viewFromAd:ad.nativeAd];
        [item.loader renderNativeAdView:templateView withAd:ad];
        item.adView = templateView;
    }

    item.creativeType = [self detectCreativeTypeFromAd:ad nativeAdView:item.adView];

    [[DemoAppLogger sharedInstance] logMessage:[NSString stringWithFormat:@"✅ Feed slot %ld loaded: %@", (long)idx, [self badgeTextForCreativeType:item.creativeType]]];

    [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:idx inSection:0]]
                          withRowAnimation:UITableViewRowAnimationFade];

    self.nextLoadIndex++;
    [self loadNextAd];
}

- (void)didFailToLoadNativeAdForAdUnitIdentifier:(NSString *)adUnitId error:(CLXError *)error {
    NSInteger idx = self.nextLoadIndex;
    if (idx >= kAdCount) return;

    NativeAdFeedItem *item = self.feedItems[idx];
    item.isLoading = NO;
    item.isFailed = YES;

    [[DemoAppLogger sharedInstance] logMessage:[NSString stringWithFormat:@"❌ Feed slot %ld failed: %@", (long)idx, error.localizedDescription]];

    [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:idx inSection:0]]
                          withRowAnimation:UITableViewRowAnimationFade];

    self.nextLoadIndex++;
    [self loadNextAd];
}

- (void)didClickNativeAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"👆 Feed didClickNativeAd" ad:ad];
}

- (void)didExpireNativeAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"⏰ Feed didExpireNativeAd" ad:ad];
}

#pragma mark - CLXAdRevenueDelegate

- (void)didPayRevenueForAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"💰 Feed didPayRevenueForAd" ad:ad];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return kAdCount;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kAdCellIdentifier forIndexPath:indexPath];

    for (UIView *subview in cell.contentView.subviews) {
        [subview removeFromSuperview];
    }

    NativeAdFeedItem *item = self.feedItems[indexPath.row];

    if (item.isLoaded && item.adView) {
        item.adView.translatesAutoresizingMaskIntoConstraints = NO;
        [cell.contentView addSubview:item.adView];

        [NSLayoutConstraint activateConstraints:@[
            [item.adView.topAnchor constraintEqualToAnchor:cell.contentView.topAnchor constant:8],
            [item.adView.leadingAnchor constraintEqualToAnchor:cell.contentView.leadingAnchor constant:12],
            [item.adView.trailingAnchor constraintEqualToAnchor:cell.contentView.trailingAnchor constant:-12],
            [item.adView.bottomAnchor constraintEqualToAnchor:cell.contentView.bottomAnchor constant:-8],
        ]];

        UILabel *badge = [[UILabel alloc] init];
        badge.text = [NSString stringWithFormat:@" %@ ", [self badgeTextForCreativeType:item.creativeType]];
        badge.font = [UIFont boldSystemFontOfSize:11];
        badge.textColor = [UIColor whiteColor];
        badge.backgroundColor = [self badgeColorForCreativeType:item.creativeType];
        badge.layer.cornerRadius = 4;
        badge.clipsToBounds = YES;
        badge.translatesAutoresizingMaskIntoConstraints = NO;
        [cell.contentView addSubview:badge];

        UILabel *slotLabel = [[UILabel alloc] init];
        slotLabel.text = [NSString stringWithFormat:@"Slot %ld", (long)indexPath.row];
        slotLabel.font = [UIFont monospacedSystemFontOfSize:10 weight:UIFontWeightMedium];
        slotLabel.textColor = [UIColor tertiaryLabelColor];
        slotLabel.translatesAutoresizingMaskIntoConstraints = NO;
        [cell.contentView addSubview:slotLabel];

        [NSLayoutConstraint activateConstraints:@[
            [badge.topAnchor constraintEqualToAnchor:cell.contentView.topAnchor constant:14],
            [badge.trailingAnchor constraintEqualToAnchor:cell.contentView.trailingAnchor constant:-18],
            [slotLabel.topAnchor constraintEqualToAnchor:cell.contentView.topAnchor constant:14],
            [slotLabel.leadingAnchor constraintEqualToAnchor:cell.contentView.leadingAnchor constant:18],
        ]];
    } else if (item.isFailed) {
        UILabel *errorLabel = [[UILabel alloc] init];
        errorLabel.text = [NSString stringWithFormat:@"Slot %ld — Load Failed", (long)indexPath.row];
        errorLabel.textColor = [UIColor systemRedColor];
        errorLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleSubheadline];
        errorLabel.textAlignment = NSTextAlignmentCenter;
        errorLabel.translatesAutoresizingMaskIntoConstraints = NO;
        [cell.contentView addSubview:errorLabel];
        [NSLayoutConstraint activateConstraints:@[
            [errorLabel.centerXAnchor constraintEqualToAnchor:cell.contentView.centerXAnchor],
            [errorLabel.centerYAnchor constraintEqualToAnchor:cell.contentView.centerYAnchor],
        ]];
    } else {
        UIActivityIndicatorView *spinner = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
        spinner.translatesAutoresizingMaskIntoConstraints = NO;
        [spinner startAnimating];
        [cell.contentView addSubview:spinner];

        UILabel *loadingLabel = [[UILabel alloc] init];
        loadingLabel.text = [NSString stringWithFormat:@"Slot %ld — Loading...", (long)indexPath.row];
        loadingLabel.textColor = [UIColor secondaryLabelColor];
        loadingLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleSubheadline];
        loadingLabel.translatesAutoresizingMaskIntoConstraints = NO;
        [cell.contentView addSubview:loadingLabel];

        [NSLayoutConstraint activateConstraints:@[
            [spinner.centerYAnchor constraintEqualToAnchor:cell.contentView.centerYAnchor],
            [spinner.trailingAnchor constraintEqualToAnchor:loadingLabel.leadingAnchor constant:-8],
            [loadingLabel.centerXAnchor constraintEqualToAnchor:cell.contentView.centerXAnchor constant:12],
            [loadingLabel.centerYAnchor constraintEqualToAnchor:cell.contentView.centerYAnchor],
        ]];
    }

    cell.backgroundColor = [UIColor clearColor];
    cell.contentView.backgroundColor = [UIColor clearColor];
    return cell;
}

#pragma mark - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    NativeAdFeedItem *item = self.feedItems[indexPath.row];

    if (item.isLoaded) {
        if (item.creativeType == NativeAdCreativeTypeVideoPortrait) {
            return 600;
        }
        return 400;
    }

    return 80;
}

@end

File: demo-app-swift/CloudXSwiftRemotePods/Views/Ad/NativeFeedViewController.swift
import UIKit
import CloudXCore

private let kAdCount = 10

private enum NativeAdCreativeType {
    case unknown, image, videoLandscape, videoPortrait

    var badgeText: String {
        switch self {
        case .unknown: return "UNKNOWN"
        case .image: return "IMAGE"
        case .videoLandscape: return "VIDEO 16:9"
        case .videoPortrait: return "REELS 9:16"
        }
    }

    var badgeColor: UIColor {
        switch self {
        case .unknown: return .systemGray
        case .image: return .systemBlue
        case .videoLandscape: return .systemOrange
        case .videoPortrait: return .systemPurple
        }
    }
}

private class NativeAdFeedItem {
    let loader: CLXNativeAdLoader
    var adView: CLXNativeAdView?
    var ad: CLXAd?
    var creativeType: NativeAdCreativeType = .unknown
    var isLoading = false
    var isLoaded = false
    var isFailed = false

    init(loader: CLXNativeAdLoader) {
        self.loader = loader
    }
}

class NativeFeedViewController: UIViewController, UITableViewDataSource, UITableViewDelegate,
                                 CLXNativeAdDelegate, CLXAdRevenueDelegate {

    private let tableView = UITableView(frame: .zero, style: .plain)
    private var feedItems: [NativeAdFeedItem] = []
    private var nextLoadIndex = 0
    private let settings = UserDefaultsSettings.shared

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Native Feed"
        view.backgroundColor = .systemBackground
        setupTableView()
        setupAppLogsButton()
        createFeedItems()
        loadNextAd()
    }

    deinit {
        feedItems.forEach { $0.loader.destroy() }
    }

    // MARK: - Setup

    private func setupTableView() {
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.dataSource = self
        tableView.delegate = self
        tableView.separatorStyle = .none
        tableView.allowsSelection = false
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "NativeAdCell")
        view.addSubview(tableView)

        NSLayoutConstraint.activate([
            tableView.topAnchor.constraint(equalTo: view.topAnchor),
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
        ])
    }

    private func setupAppLogsButton() {
        let logsButton = UIButton(type: .system)
        logsButton.setTitle("App Logs", for: .normal)
        logsButton.titleLabel?.font = .systemFont(ofSize: 14, weight: .medium)
        logsButton.backgroundColor = .systemOrange
        logsButton.setTitleColor(.white, for: .normal)
        logsButton.layer.cornerRadius = 6
        logsButton.translatesAutoresizingMaskIntoConstraints = false
        logsButton.addTarget(self, action: #selector(showLogsModal), for: .touchUpInside)
        view.addSubview(logsButton)

        NSLayoutConstraint.activate([
            logsButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 8),
            logsButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -12),
            logsButton.widthAnchor.constraint(equalToConstant: 90),
            logsButton.heightAnchor.constraint(equalToConstant: 30),
        ])
    }

    @objc private func showLogsModal() {
        let logsModal = LogsModalViewController(title: "Logs")
        present(logsModal, animated: true)
    }

    private var adUnitId: String {
        let id = CLXDemoConfigManager.sharedManager.currentConfig.nativeAdUnitId
        return settings.nativeMediumAdUnitId.isEmpty ? id : settings.nativeMediumAdUnitId
    }

    private func createFeedItems() {
        for i in 0..<kAdCount {
            let loader = CloudXCore.shared.createNativeAdLoader(adUnitIdentifier: adUnitId)
            loader.nativeAdDelegate = self
            loader.revenueDelegate = self
            loader.placement = "feed_slot_\(i)"
            feedItems.append(NativeAdFeedItem(loader: loader))
        }
    }

    // MARK: - Sequential Loading

    private func loadNextAd() {
        guard nextLoadIndex < kAdCount else { return }
        let item = feedItems[nextLoadIndex]
        item.isLoading = true
        item.loader.loadAd()
    }

    // MARK: - Creative Type Detection

    private func detectCreativeType(from ad: CLXAd) -> NativeAdCreativeType {
        guard let nativeAd = ad.nativeAd else { return .unknown }

        if nativeAd.isVideoContent {
            let ar = nativeAd.mediaContentAspectRatio
            return ar < 1.0 ? .videoPortrait : .videoLandscape
        }

        return nativeAd.mediaContentAspectRatio > 0 ? .image : .unknown
    }

    // MARK: - CLXNativeAdDelegate

    func didLoadNativeAd(_ nativeAdView: CLXNativeAdView?, for ad: CLXAd) {
        let idx = nextLoadIndex
        guard idx < kAdCount else { return }

        let item = feedItems[idx]
        item.isLoading = false
        item.isLoaded = true
        item.ad = ad

        if let view = nativeAdView {
            item.adView = view
        } else if let nativeAd = ad.nativeAd {
            let templateView = CLXNativeAdView(from: nativeAd)
            item.loader.renderNativeAdView(templateView, with: ad)
            item.adView = templateView
        }

        item.creativeType = detectCreativeType(from: ad)
        DemoAppLogger.sharedInstance.logMessage("✅ Feed slot \(idx) loaded: \(item.creativeType.badgeText)")

        tableView.reloadRows(at: [IndexPath(row: idx, section: 0)], with: .fade)

        nextLoadIndex += 1
        loadNextAd()
    }

    func didFailToLoadNativeAd(forAdUnitIdentifier adUnitId: String, error: CLXError) {
        let idx = nextLoadIndex
        guard idx < kAdCount else { return }

        let item = feedItems[idx]
        item.isLoading = false
        item.isFailed = true

        DemoAppLogger.sharedInstance.logMessage("❌ Feed slot \(idx) failed: \(error.localizedDescription)")

        tableView.reloadRows(at: [IndexPath(row: idx, section: 0)], with: .fade)

        nextLoadIndex += 1
        loadNextAd()
    }

    func didClickNativeAd(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("👆 Feed didClickNativeAd", ad: ad)
    }

    func didExpireNativeAd(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("⏰ Feed didExpireNativeAd", ad: ad)
    }

    // MARK: - CLXAdRevenueDelegate

    func didPayRevenue(for ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("💰 Feed didPayRevenueForAd", ad: ad)
    }

    // MARK: - UITableViewDataSource

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int { kAdCount }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "NativeAdCell", for: indexPath)
        cell.contentView.subviews.forEach { $0.removeFromSuperview() }

        let item = feedItems[indexPath.row]

        if item.isLoaded, let adView = item.adView {
            adView.translatesAutoresizingMaskIntoConstraints = false
            cell.contentView.addSubview(adView)
            NSLayoutConstraint.activate([
                adView.topAnchor.constraint(equalTo: cell.contentView.topAnchor, constant: 8),
                adView.leadingAnchor.constraint(equalTo: cell.contentView.leadingAnchor, constant: 12),
                adView.trailingAnchor.constraint(equalTo: cell.contentView.trailingAnchor, constant: -12),
                adView.bottomAnchor.constraint(equalTo: cell.contentView.bottomAnchor, constant: -8),
            ])

            let badge = UILabel()
            badge.text = " \(item.creativeType.badgeText) "
            badge.font = .boldSystemFont(ofSize: 11)
            badge.textColor = .white
            badge.backgroundColor = item.creativeType.badgeColor
            badge.layer.cornerRadius = 4
            badge.clipsToBounds = true
            badge.translatesAutoresizingMaskIntoConstraints = false
            cell.contentView.addSubview(badge)

            let slotLabel = UILabel()
            slotLabel.text = "Slot \(indexPath.row)"
            slotLabel.font = .monospacedSystemFont(ofSize: 10, weight: .medium)
            slotLabel.textColor = .tertiaryLabel
            slotLabel.translatesAutoresizingMaskIntoConstraints = false
            cell.contentView.addSubview(slotLabel)

            NSLayoutConstraint.activate([
                badge.topAnchor.constraint(equalTo: cell.contentView.topAnchor, constant: 14),
                badge.trailingAnchor.constraint(equalTo: cell.contentView.trailingAnchor, constant: -18),
                slotLabel.topAnchor.constraint(equalTo: cell.contentView.topAnchor, constant: 14),
                slotLabel.leadingAnchor.constraint(equalTo: cell.contentView.leadingAnchor, constant: 18),
            ])
        } else if item.isFailed {
            let errorLabel = UILabel()
            errorLabel.text = "Slot \(indexPath.row) — Load Failed"
            errorLabel.textColor = .systemRed
            errorLabel.font = .preferredFont(forTextStyle: .subheadline)
            errorLabel.textAlignment = .center
            errorLabel.translatesAutoresizingMaskIntoConstraints = false
            cell.contentView.addSubview(errorLabel)
            NSLayoutConstraint.activate([
                errorLabel.centerXAnchor.constraint(equalTo: cell.contentView.centerXAnchor),
                errorLabel.centerYAnchor.constraint(equalTo: cell.contentView.centerYAnchor),
            ])
        } else {
            let spinner = UIActivityIndicatorView(style: .medium)
            spinner.translatesAutoresizingMaskIntoConstraints = false
            spinner.startAnimating()
            cell.contentView.addSubview(spinner)

            let loadingLabel = UILabel()
            loadingLabel.text = "Slot \(indexPath.row) — Loading..."
            loadingLabel.textColor = .secondaryLabel
            loadingLabel.font = .preferredFont(forTextStyle: .subheadline)
            loadingLabel.translatesAutoresizingMaskIntoConstraints = false
            cell.contentView.addSubview(loadingLabel)

            NSLayoutConstraint.activate([
                spinner.centerYAnchor.constraint(equalTo: cell.contentView.centerYAnchor),
                spinner.trailingAnchor.constraint(equalTo: loadingLabel.leadingAnchor, constant: -8),
                loadingLabel.centerXAnchor.constraint(equalTo: cell.contentView.centerXAnchor, constant: 12),
                loadingLabel.centerYAnchor.constraint(equalTo: cell.contentView.centerYAnchor),
            ])
        }

        cell.backgroundColor = .clear
        cell.contentView.backgroundColor = .clear
        return cell
    }

    // MARK: - UITableViewDelegate

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let item = feedItems[indexPath.row]
        if item.isLoaded {
            return item.creativeType == .videoPortrait ? 600 : 400
        }
        return 80
    }
}
